compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -744,6 +744,52 @@ impl<'tcx> PolyExistentialPredicate<'tcx> {
 }
 
 impl<'tcx> List<ty::PolyExistentialPredicate<'tcx>> {
+    /// Returns the list of arguments if principal is an Fn trait.
+    pub fn fn_trait_args(&self, tcx: TyCtxt<'tcx>) -> Option<&List<Ty<'tcx>>> {
+        if self.is_fn_trait(tcx) {
+            if let Some(args) = self.principal_args() && !args.is_empty() && args[0].expect_ty().is_tuple() {
+                Some(args[0].expect_ty().tuple_fields())
+            } else {
+                Some(List::empty())
```

Comment:
I changed it to a `bug!`, but interestingly I found non tuple args in an Fn trait when rebuilding the standard library:

```
...
error: internal compiler error: compiler/rustc_middle/src/ty/sty.rs:753:17: fn_trait_args: unexpected non tuple arg `Some([&ReEarlyBound(DefId(0:52682 ~ core[e0f7]::ops::function::impls::{impl#3}::'_), 0, '_) mut F/#2, A/#1])`
...
```

From the backtrace it seems it's because we also encode type ids earlier before monomorphization/substitution and update these later:

```
  24:     0x7f2941b04411 - rustc_symbol_mangling[a3dfc9c41c10ddbf]::typeid::typeid_for_instance
  25:     0x7f293cfd1c97 - <rustc_codegen_llvm[4e537aa28e437798]::context::CodegenCx>::declare_fn
  26:     0x7f293cfd284f - <rustc_codegen_llvm[4e537aa28e437798]::context::CodegenCx as rustc_codegen_ssa[d28a52787be8e29]::traits::declare::PreDefineMethods>::predefine_fn
```
Do you have any suggestions of what to do here?

Reply:
Why are we not monomorphizing type ids? This means that polymorphic fns totally don't include the right args.

Reply:
Also, where are these arguments coming from? Why is there a lifetime in the trait args of an `Fn` existential trait ref?

Reply:
> Why are we not monomorphizing type ids? This means that polymorphic fns totally don't include the right args.

I think I'm actually wrong about it and these encoded type ids after monomorphization/substitution--see below.

> Also, where are these arguments coming from? Why is there a lifetime in the trait args of an Fn existential trait ref?

I figured it out. The problem was when handling the actual Fn traits default implementations and replacing the concrete self by a reference to a trait object--this is the only place it could have happened.

Fn traits are not transformed into their identities because they are transformed into function pointers before (and should never occur in transform_predicate) so now I remove the unsubstituted args from their trait objects when replacing the concrete self by a reference to a trait object for the actual Fn traits default implementations.